src/xcode/ENA/ENA/Source/Services/RecycleBin/RecycleBin.swift

```diff
@@ -0,0 +1,207 @@
+	///
+	/// - Parameter item: The item which needs to be moved into the bin.
+	/// - Parameter recycledAt: Date at which the item was moved into the bin. Default value: `Date()`
+	/// - Returns: A `RecycleBinItem` which contains the item moved into the bin and metadata added from the bin (like `recycledAt`).
+	@discardableResult
+	func moveToBin(_ item: RecycledItem, recycledAt: Date = Date()) -> RecycleBinItem {
+		Log.info("Recycle item.", log: .recycleBin)
+
+		let binItem = RecycleBinItem(
+			recycledAt: recycledAt,
+			item: item
+		)
+		store.recycleBinItems.insert(binItem)
+		return binItem
+	}
+
+	/// Asks if an item can be restored. This call is not restoring the item immediately.
+	/// Instead its asking the restoration handler, if the item can be restored.
+	///	In some cases the item cannot be restored immediately and needs some further processing or a user decision.
+	///	Please see `restore(...)` for restoring the item immediately (Call it after further processing was done and the item can finally be restored).
+	///
+	/// - Parameter item: The item which needs to be restored.
+	/// - Returns: `Result.Success`: The item can be restored. `Result.Failure`: Item cannot be restored and needs some further processing.
+	func canRestore(_ item: RecycleBinItem) -> Result<Void, RestorationError> {
+		Log.info("Ask for item restoration.", log: .recycleBin)
+
+		switch item.item {
+		case .certificate(let certificate):
+			Log.info("Ask for certificate item restoration.", log: .recycleBin)
+			let canRestoreResult = certificateRestorationHandler.canRestore(certificate)
+			return canRestoreResult.mapError { RestorationError.certificateError($0) }
+		case .coronaTest(let coronaTest):
+			Log.info("Ask for test item restoration.", log: .recycleBin)
+			let canRestoreResult = testRestorationHandler.canRestore(coronaTest)
+			return canRestoreResult.mapError { RestorationError.testError($0) }
+		}
+	}
+
+	/// Restores an item.
+	/// Restoration handler is called to restore the item.
+	/// The item is removed from the bin.
+	///
+	/// - Parameter item: The item which needs to be restored.
+	func restore(_ item: RecycleBinItem) {
+		Log.info("Restore item.", log: .recycleBin)
+
+		switch item.item {
+		case .certificate(let certificate):
+			Log.info("Restore certificate item.", log: .recycleBin)
+			certificateRestorationHandler.restore(certificate)
+		case .coronaTest(let coronaTest):
+			Log.info("Restore test item.", log: .recycleBin)
+			testRestorationHandler.restore(coronaTest)
+		}
+
+		store.recycleBinItems.remove(item)
+	}
+
+	/// Removes the item from the bin.
+	///
+	/// - Parameter item: The item which needs to be removed.
+	func remove(_ item: RecycleBinItem) {
+		Log.info("Remove item.", log: .recycleBin)
+		store.recycleBinItems.remove(item)
+	}
+
+	/// Removes all items from the bin.
+	func removeAll() {
+		Log.info("Remove all items.", log: .recycleBin)
+		store.recycleBinItems.removeAll()
+	}
```

Comment:
Does it make sense to have a func `removeAllCertificates` or `removeAllTests`?

Reply:
At the moment not, because from UI-perspective there won't be an option to delete the item-types separately.
Maybe it will come later, then we still can add it additionally.